### DISCUSSION TITLE
Add additional metrics reported by Zeek

### DISCUSSION THREAD

**Discussion by timwoj** [zeek/zeek#1613] [Ideas]
With the new Telemetry feature added recently, we can now define additional metrics and telemetry data that gets reported externally. This discussion is to build a list of useful metrics to report. I've made it a checklist in case people want to start working on them, they can check them off as they're added.

- [ ] Active connections/sessions
- [ ] Load on event loop (@ckreibich, can you expound on what you meant here?)
- [ ] Message queue statistics

**Comment by grigorescu** [comment]
I was chatting with Seth about this a bit. I'm intrigued by a new attribute `&track_size` or something, which can be applied to tables. It would hook into adding and removing values, in the core, and update a metric in the Telemetry framework. For size, I just mean number of elements and not memory usage.

I think it'd be especially interesting because I can `redef` this on a table from a script in base, in case I want to track that.

**Reply by JustinAzoff** [reply]
Would that only be for global tables, or could it somehow also track lots of small tables that are added to something like a connection object?

**Reply by rsmmr** [reply]
Whatever tables we can get attributes on, I image (potentially via the `local` trick). However, as always with tables, we wouldn't be able to count across nested tables (i.e., a table of table where an inner one is modified wouldn't show up for the size tracking of  the outer).

**Comment by JustinAzoff** [comment]
For current stats see what we log using the misc/stats.zeek script.  one can `redef Stats::report_interval=30secs;` to get them faster, but having them be done via metrics would be even better.

New things that would be nice:

- [ ] Broker event counters - if possible on a per event type basis.

**Comment by rsmmr** [comment]
- [x] cluster per-node CPU
- [x] cluster per-node memory
- [ ] cluster per-node message volume incoming / outgoing

They are probably included in the above notes one way or the other already, but just to call them out.

**Reply by Neverlord** [reply]
CPU and memory are usually collected by CAF, but they're currently missing. See: https://github.com/zeek/broker/issues/189.

**Comment by ckreibich** [comment]
My comment on the event loop load was to understand how many events we process per queue drain. We have the `num_events_dispatched` member in the event manager and the `EventStats` record. Both are related. It'd be a bit like events per time unit, but more tied to the implementation. I'm mainly just curious about this one and events per time unit is probably fine.

That was almost an aside, though. In general I'd really like to see more metrics that track cluster load. I may not have sufficient understanding here (particularly with CAF's credit system) but it'd be cool to see how many messages CAF receives in each of its polls. Is that the same as `num_pending_queries` in `BrokerStats`? What do we need to track to understand whether a node can keep up with incoming messages? It'd be good here to have message sender name in the metric, though that could add many dozens of metric names for large clusters. (A gauge for time from message arrival to message processing would be related — not sure that's feasible?)

Another thing that'd be really handy, but I don't know feasibility: visibility into the most frequent hits on a pool of strings/names per time unit. Picture our event names — what's are the most busy events on a given node? I guess we could approximate via one metric per event type, but that feels unwieldy.
